perception/sub8_perception/src/sub8_perception/navigation_gate.cpp

```diff
@@ -0,0 +1,319 @@
+  {
+    // Check if the contour is a gate
+    if (valid_contour(contour.at(i)))
+    {
+      // Match the contour with a polygon
+      auto epsilon = 0.01 * cv::arcLength(contour.at(i), true);
+      cv::approxPolyDP(contour.at(i), features, epsilon, true);
+    }
+  }
+  return features;
+}
+
+void Sub8NavigationGateDetector::visualize_k_gate_normal()
+{
+  geometry_msgs::Point center_pt;
+  center_pt.x = gate_pose_.translation().x();
+  center_pt.y = gate_pose_.translation().y();
+  center_pt.z = gate_pose_.translation().z();
+
+  Eigen::Vector3d normal = gate_pose_.rotation() * Eigen::Vector3d(1, 0, 0);
+
+  geometry_msgs::Point sdp_normalvec_ros;
+  sdp_normalvec_ros.x = center_pt.x + normal(0, 0);
+  sdp_normalvec_ros.y = center_pt.y + normal(1, 0);
+  sdp_normalvec_ros.z = center_pt.z + normal(2, 0);
+
+  visualization_msgs::Marker marker_normal;
+  marker_normal.header.seq = 0;
+  marker_normal.header.stamp = ros::Time::now();
+  marker_normal.header.frame_id = "/front_left_cam";
+  marker_normal.id = 2222;
+  marker_normal.type = visualization_msgs::Marker::ARROW;
+  marker_normal.points.push_back(center_pt);
+  marker_normal.points.push_back(sdp_normalvec_ros);
+  marker_normal.scale.x = 0.1;
+  marker_normal.scale.y = 0.5;
+  marker_normal.scale.z = 0.5;
+  marker_normal.color.a = 1.0;
+  marker_normal.color.r = 1.0;
+  marker_normal.color.g = 0.0;
+  marker_normal.color.b = 1.0;
+  marker_pub_.publish(marker_normal);
+}
+
+void Sub8NavigationGateDetector::visualize_3d_points_rviz(const std::vector<Eigen::Vector3d> &feature_pts_3d)
+{
+  visualization_msgs::Marker points_raw_marker;
+  points_raw_marker.header.stamp = ros::Time::now();
+  points_raw_marker.type = visualization_msgs::Marker::POINTS;
+  points_raw_marker.header.frame_id = "/front_left_cam";
+
+  points_raw_marker.id = 5;
+  points_raw_marker.scale.x = 0.2;
+  points_raw_marker.scale.y = 0.2;
+  points_raw_marker.color.r = 1.0f;
+  points_raw_marker.color.a = 1.0;
+
+  for (size_t i = 0; i < feature_pts_3d.size(); i++)
+  {
+    geometry_msgs::Point p;
+    p.x = feature_pts_3d[i][0];
+    p.y = feature_pts_3d[i][1];
+    p.z = feature_pts_3d[i][2];
+    points_raw_marker.points.push_back(p);
+  }
+
+  marker_pub_.publish(points_raw_marker);
+}
+
+std::vector<cv::Point> Sub8NavigationGateDetector::get_corner_center_points(const std::vector<cv::Point> &features)
+{
+  // Get all the possible number of combinations of points
+  std::vector<std::vector<uint8_t>> id_comb;
+  mil_tools::combinations(features.size(), 2, id_comb);
+
+  // Order pairs of points based of shortest distance. store that in id_comb
+  std::sort(id_comb.begin(), id_comb.end(),
+            [&features](const std::vector<uint8_t> &a, const std::vector<uint8_t> &b) -> bool {
+              auto diffx = features[a[0]].x - features[a[1]].x;
+              auto diffy = features[a[0]].y - features[a[1]].y;
+              auto distanceA = std::sqrt(std::pow(diffx, 2) + std::pow(diffy, 2));
```

Comment:
Consider using `std::hypot(diffx, diffy)` to avoid loss of precision

Also: Doesn't this have wacky behavior if you have more than 256 features?